### DISCUSSION TITLE
filter by  $__timeFilter using jsonata

### DISCUSSION THREAD

**Discussion by yosiasz** [grafana/grafana-infinity-datasource#315] [Q&A] [answered]
Greetings given the following sample data

```
[
	{
		"Name": "Zoro",
		"Place": "New York",
		"date": 1355529600,
		"Longitude": 40.6943,
		"Latitude": -73.9249
	},
	{
		"Name": "Vader",
		"Place": "Beijing",
		"date": 1355184000,
		"Longitude": 39.904,
		"Latitude": 116.4075
	}
]
```

How can one filter by the date range selected?
```
parse-json
| project "Name", "Place","dt"=unixtime_seconds_todatetime("date"),"Longitude","Latitude"
| jsonata { 'dt': $__timeFilter} 
| order by "dt" asc
```

Thanks

**Comment by yesoreyeram** [comment, marked as answer]
@yosiasz - `$__timeFilter` is not a macro in infinity. Typically, SQL datasources in grafana uses that macro and they are aware how to expand that. But infinity is a generic plugin and doesn't aware how to expand that macro.

In your case, I would prefer something like [this](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fjsonplaceholder.typicode.com%2Fusers%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22data%22:%22%5B%5Cn%5Ct%7B%5Cn%5Ct%5Ct%5C%22Name%5C%22:%20%5C%22Zoro%5C%22,%5Cn%5Ct%5Ct%5C%22Place%5C%22:%20%5C%22New%20York%5C%22,%5Cn%5Ct%5Ct%5C%22date%5C%22:%201355529600,%5Cn%5Ct%5Ct%5C%22Longitude%5C%22:%2040.6943,%5Cn%5Ct%5Ct%5C%22Latitude%5C%22:%20-73.9249%5Cn%5Ct%7D,%5Cn%5Ct%7B%5Cn%5Ct%5Ct%5C%22Name%5C%22:%20%5C%22Vader%5C%22,%5Cn%5Ct%5Ct%5C%22Place%5C%22:%20%5C%22Beijing%5C%22,%5Cn%5Ct%5Ct%5C%22date%5C%22:%201355184000,%5Cn%5Ct%5Ct%5C%22Longitude%5C%22:%2039.904,%5Cn%5Ct%5Ct%5C%22Latitude%5C%22:%20116.4075%5Cn%5Ct%7D%5Cn%5D%22,%22uql%22:%22parse-json%5Cn%7C%20jsonata%20%5C%22*%5Bdate%3E%3D$%7B__from:date:seconds%7D%20and%20date%3C%3D$%7B__to:date:seconds%7D%5D%5C%22%5Cn%7C%20extend%20%5C%22date%5C%22%3Dunixtime_seconds_todatetime(%5C%22date%5C%22)%5Cn%7C%20order%20by%20%5C%22date%5C%22%20asc%22%7D%5D,%22range%22:%7B%22from%22:%22now-25y%22,%22to%22:%22now%22%7D%7D)

```kusto
parse-json
| jsonata "*[date>=${__from:date:seconds} and date<=${__to:date:seconds}]"
| extend "date"=unixtime_seconds_todatetime("date")
| order by "date" asc
```

Let me know if this solves your issue or if you find any bug with the implementation.

**Reply by yosiasz** [reply]
Looks wonderful except when you use certain specific date ranges such 2012-12-14 00:00:00, the result set is as follows
```
{ "Name": "Zoro", "Place": "New York", "Longitude": 40.6943, "Latitude": -73.9249 }
```

no longer a table. interesting.

**Reply by yesoreyeram** [reply]
created an issue for this https://github.com/yesoreyeram/uql/issues/5

**Reply by gdiana** [reply]
The problem might lie in your date format. The date format you are using is in seconds and not milliseconds. UQL and jsonata uses ISO 8601 and will otherwise throw an error or just omit if using jsonata.

When you use date in jsonata you must use 2012-12-14T00:00:00 and not 2012-12-14 00:00:00

![image](https://user-images.githubusercontent.com/13445152/168796552-5ed523b6-78ad-4bdf-8a20-da0878d5056f.png)

**Reply by yosiasz** [reply]
wonder what would happen if the result set is just one. testing

**Reply by gdiana** [reply]
This works multiple or single

parse-json
| jsonata "*.{ 'time' : $fromMillis(date.$number()*1000),'Name' : Name,'Place' : Place,'Lat': Latitude,'Long' : Longitude}"

It just reformulates the json and converts date to IOS format



![image](https://user-images.githubusercontent.com/13445152/169065585-26640138-ced2-4634-ab31-25e3974b67ac.png)

**Reply by yesoreyeram** [reply]
Glad it works for you @gdiana . The original question was about using grafana dashboard time range filter here

**Reply by yosiasz** [reply]
Got it confused with the other issue 😆 thanks!
